File: grain/proxy/forward.py
"""Forwards a validated request to the real GitHub smart-HTTP endpoint.

Behind a `Forwarder` interface for the same reason `run.py` puts command
execution behind a `Runner`: the proxy's decision logic (core.py) is worth
testing without a real network call to github.com on every test run.
"""

from __future__ import annotations

import base64
import http.client
from dataclasses import dataclass, field
from typing import Protocol


@dataclass(frozen=True)
class UpstreamResponse:
    status: int
    headers: dict[str, str]
    body: bytes


class Forwarder(Protocol):
    def forward(self, *, method: str, path: str, query: str,
                headers: dict[str, str], body: bytes | None,
                token: str | None) -> UpstreamResponse: ...


class RealForwarder:
    """Talks to GitHub's smart-HTTP endpoint over HTTPS by default.
    `use_tls=False` is the git-transport half of the same mock-GitHub test
    seam as `RealTransport` (`grain/automation/github.py`) — see
    `grain/proxy/server.py`'s `build_proxy`, which wires both from the same
    `automation.json` override so a live test can point clone/push and the
    REST API at one mock server together.
    """

    def __init__(self, host: str = "github.com", *, use_tls: bool = True) -> None:
        self.host = host
        self.use_tls = use_tls

    def forward(self, *, method: str, path: str, query: str,
                headers: dict[str, str], body: bytes | None,
                token: str | None) -> UpstreamResponse:
        out_headers = {
            "Content-Type": headers.get("Content-Type", ""),
            "Accept": headers.get("Accept", "*/*"),
            "User-Agent": headers.get("User-Agent", "git/grain-proxy"),
        }
        if token:
            # GitHub's HTTPS PAT convention: the token as the Basic auth
            # username, empty password.
            out_headers["Authorization"] = (
                "Basic " + base64.b64encode(f"{token}:".encode()).decode()
            )
        out_headers = {k: v for k, v in out_headers.items() if v}

        conn_cls = http.client.HTTPSConnection if self.use_tls else http.client.HTTPConnection
        conn = conn_cls(self.host, timeout=30)
        try:
            full_path = f"{path}?{query}" if query else path
            conn.request(method, full_path, body=body, headers=out_headers)
            resp = conn.getresponse()
            data = resp.read()
            return UpstreamResponse(resp.status, dict(resp.getheaders()), data)
        finally:
            conn.close()


@dataclass
class FakeForwarder:
    """Records calls and replays a scripted response. For tests."""

    response: UpstreamResponse = field(
        default_factory=lambda: UpstreamResponse(200, {}, b"")
    )
    calls: list[dict] = field(default_factory=list)

    def forward(self, *, method: str, path: str, query: str,
                headers: dict[str, str], body: bytes | None,
                token: str | None) -> UpstreamResponse:
        self.calls.append(
            {"method": method, "path": path, "query": query,
             "headers": dict(headers), "body": body, "token": token}
        )
        return self.response

File: v2/gitproxy/forward.go
package gitproxy

// Forwards a validated request to the real GitHub smart-HTTP endpoint.
//
// Behind a Forwarder interface for the same reason command execution sits
// behind an interface elsewhere in this project: the proxy's decision
// logic (core.go) is worth testing without a real network call to
// github.com on every test run.

import (
	"bytes"
	"encoding/base64"
	"io"
	"net/http"
)

// UpstreamResponse is what the forward target answered.
type UpstreamResponse struct {
	Status  int
	Headers map[string]string
	Body    []byte
}

// Forwarder sends one validated request on to the real git host.
type Forwarder interface {
	Forward(method, path, query string, headers map[string]string, body []byte, token *string) (UpstreamResponse, error)
}

// RealForwarder talks to a git smart-HTTP endpoint over HTTPS by default.
// UseTLS false is the git-transport half of the mock-GitHub test seam:
// it lets a live test point a real GitProxy's forward target at a local
// stand-in server instead of github.com.
type RealForwarder struct {
	Host   string
	UseTLS bool
	Client *http.Client
}

// NewRealForwarder returns a forwarder aimed at host over HTTPS.
func NewRealForwarder(host string) *RealForwarder {
	return &RealForwarder{Host: host, UseTLS: true}
}

func (f *RealForwarder) Forward(method, path, query string, headers map[string]string, body []byte, token *string) (UpstreamResponse, error) {
	scheme := "https"
	if !f.UseTLS {
		scheme = "http"
	}
	url := scheme + "://" + f.Host + path
	if query != "" {
		url += "?" + query
	}

	var bodyReader io.Reader
	if body != nil {
		bodyReader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, url, bodyReader)
	if err != nil {
		return UpstreamResponse{}, err
	}

	setHeader(req, "Content-Type", headers["Content-Type"])
	accept := headers["Accept"]
	if accept == "" {
		accept = "*/*"
	}
	setHeader(req, "Accept", accept)
	userAgent := headers["User-Agent"]
	if userAgent == "" {
		userAgent = "git/grain-proxy"
	}
	setHeader(req, "User-Agent", userAgent)
	// git gzip-compresses a git-upload-pack request body once its
	// "have"/"want" negotiation gets large enough (http.postBuffer) --
	// forwarding a compressed body without this header means the upstream
	// tries to parse gzip bytes as plain pkt-line data.
	setHeader(req, "Content-Encoding", headers["Content-Encoding"])
	if token != nil {
		// GitHub's HTTPS PAT convention: the token as the Basic auth
		// username, empty password.
		req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(*token+":")))
	}

	client := f.Client
	if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return UpstreamResponse{}, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return UpstreamResponse{}, err
	}
	respHeaders := map[string]string{}
	for k := range resp.Header {
		respHeaders[k] = resp.Header.Get(k)
	}
	return UpstreamResponse{Status: resp.StatusCode, Headers: respHeaders, Body: data}, nil
}

func setHeader(req *http.Request, key, value string) {
	if value != "" {
		req.Header.Set(key, value)
	}
}

// FakeForwarder records calls and replays a scripted response. For tests.
type FakeForwarder struct {
	Response UpstreamResponse
	Err      error
	Calls    []ForwardCall
}

// ForwardCall is one recorded call to FakeForwarder.Forward.
type ForwardCall struct {
	Method  string
	Path    string
	Query   string
	Headers map[string]string
	Body    []byte
	Token   *string
}

func (f *FakeForwarder) Forward(method, path, query string, headers map[string]string, body []byte, token *string) (UpstreamResponse, error) {
	f.Calls = append(f.Calls, ForwardCall{Method: method, Path: path, Query: query, Headers: headers, Body: body, Token: token})
	if f.Err != nil {
		return UpstreamResponse{}, f.Err
	}
	return f.Response, nil
}
